src/lib.rs

```diff
@@ -336,25 +378,41 @@ fn handle_packet<H, W>(data: &[u8],
         match command {
             Command::EnableExtendedMode => Response::Empty,
             Command::TargetHaltReason => Response::Empty,
-            Command::ToggleDebug => Response::Empty,
             Command::ReadGeneralRegisters => Response::Empty,
             // The k packet requires no response.
             Command::Kill(None) => Response::Empty,
             // We don't implement this, so return an error.
-            Command::Kill(Some(_)) => Response::String("E01"),
+            Command::Kill(Some(_)) => Response::Error(NOT_IMPLEMENTED),
             Command::Reset => Response::Empty,
-            Command::ReadRegister(_) | Command::MemoryRead(_, _) => {
-                // We don't implement this, so return an error.
-                Response::String("E01")
+            Command::ReadRegister(regno) => {
+                match handler.read_register(regno) {
+                    Result::Ok(bytes) => Response::Bytes(bytes),
+                    Result::Err(SimpleError::Error(val)) => Response::Error(val),
```

Comment:
Is there a nice way that we can refactor the `match` handling on `Handler` results?  We're going to be writing a lot of these, and it'd be nice if we could just say:

```rust
handler.read_register(regno).to_response()
```

(or `to_response(handler.read_register(...))`) and have that take care of all of the necessary conversions for us.  I know how to do it in C++, but I don't know of a good meta-programming way in Rust to go from the success type in a `Result` and deliver the corresponding `Response` enum.

Reply:
We should be able to do something like:
```rust
impl<T> Into<Response> for Result<T, SimpleError>
  where T: Into<Response>,
{
  fn into(self) -> Response {
    match self {
      Err(SimpleError::Error(val)) => Reponse::Error(val),
      Ok(t) => t.into(),
    }
 }
}
```

(obviously we'd also need to implement `Into<Response>` for whatever other types we want to handle there). Then you could just write `handler.read_register(regno).into()`.

Related: we might want to pull in something like `failure` to define an error type and get automatic conversion.

Reply:
I'll send a PR shortly to implement `From`